### PR TITLE
net: lantiq: backport latest upstream patches 

### PR DESCRIPTION
**Pull request by abajk**
This patch includes a series of performance improvements. All patches should land in 5.17. This PR should be merged after https://github.com/openwrt/openwrt/pull/4353.
NAT Performance results on BT Home Hub 5A (kernel 5.10.89, mtu 1500):

~~~
	Down		Up
Before	539 Mbps	599 Mbps
After	624 Mbps	695 Mbps
~~~

Signed-off-by: Aleksander Jan Bajkowski <olek2@wp.pl>

### CODE REVIEW

**Comment by jospezial** [comment]
> This PR should be merged after #4353.

In this PR here you already included 
lantiq: xrx200: enable use of baby jumbo frames
https://github.com/openwrt/openwrt/pull/4939/commits/872a17d6688fb044097520d30d45fb64fc2c5d2d
from #4353
Can #4353 be closed then?

Btw, I have seen
https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/commit/?h=linux-5.10.y&id=8d70dc0eecf0fb1f1d6980fd525243afb69e3e0e
"net: lantiq_xrx200: fix statistics of received bytes"
is now in Linux 5.10.90 and with that in Openwrt master. ;-)

What about the other lantiq commits in
https://git.kernel.org/pub/scm/linux/kernel/git/netdev/net-next.git/log/?qt=grep&q=lantiq
(mostly lantiq_etop commits)?
Would it be good to add them to openwrt now too?
It is is hard for me to keep an overview on what we still have in openwrt with latest 5.10 kernel release or patches.

**Comment by abajk** [comment]
> In this PR here you already included
> lantiq: xrx200: enable use of baby jumbo frames
> [872a17d](https://github.com/openwrt/openwrt/commit/872a17d6688fb044097520d30d45fb64fc2c5d2d)
> from #4353
> Can #4353 be closed then?

The patch that converts to build_skb depends on the patches that add jumbo frames. This is the reason why the changes from #4353 are included here. I think PR 4353 shouldn't be closed. Patch written by Thomas still needs to be discussed upstram.

> What about the other lantiq commits in
> https://git.kernel.org/pub/scm/linux/kernel/git/netdev/net-next.git/log/?qt=grep&q=lantiq
> (mostly lantiq_etop commits)?
> Would it be good to add them to openwrt now too?
> It is is hard for me to keep an overview on what we have yet in openwrt with latest 5.10 kernel release or patches.

These few patches do not fix any bugs. They only contain minor fixes reported by checkpatch. I don't think it's profitable to backport them to OpenWRT now.

**Comment by nicefile** [comment]
Installed and tested on FB 7362sl 

Ports 1 ,2 on internal phy work ok
Ports 3,4 on external 10/100 phy droping pings and laggy (not related to this pull)
`root@7412:~# fping -M -p 1474 -s -q -c 10 192.168.1.1
192.168.1.1 : xmt/rcv/%loss = 10/7/30%, min/avg/max = 0.576/89.6/624

       1 targets
       1 alive
       0 unreachable
       0 unknown addresses

       3 timeouts (waiting for response)
      10 ICMP Echos sent
       8 ICMP Echo Replies received
       0 other ICMP received

 0.576 ms (min round trip time)
 89.6 ms (avg round trip time)
 624 ms (max round trip time)
       14.743 sec (elapsed real time)

`

**Comment by hauke** [comment]
Thank you for your patches, I applied them to master in 255268ce1a218a670e653dc5c83067f704164b7c and 7e484b902d269d8fd22afeccb765ccc2cd8efa19.

**Comment by jospezial** [comment]
> @abajk @hauke @kestrel1974 When trying to implement mdio writes for bringing up Fritzbox devices secondary SoC, I had to use the gswip driver mdio bus, because the access to already aquired io addresses does not work. This caused a lot of problems, because I received strange and unexpected values e.g. during mdio_read where I did not know, where they came from and initially sometimes I got around them by using a lot of mdelay() operations.
> 
> Later I found out, that phys using gswip_mdio_xxx function exposed to via mii_bus by gswip_mdio() function, use a mutex on the mii_bus->mdio_lock (See e.g. the phy_modify function in drivers/net/phy/phy-core.c). So I implemented the mutex around mdio operations and the issues went away. But any of the gswip_mdio_xxx operations inside lantiq_gswip.c do not use the same mutex on the mdio_bus lock.
> 
> I wonder if that causes or can cause any bugs, if a phy and a gswip_mdio_xxx function write/read to the mdio registers at the same time. Could that cause any bugs, race conditions or strange behaviour?
> 
> Because the gswip driver shares the mii_bus read write functions with the phys, I think lantiq_gswip.c needs to implement/use the mutex on struct mii_bus->mdio_lock for its write and read operations to the mdio registers.

@kestrel1974 
Do the lantiq commits in latest kernel releases 5.10.101, 5.10.102 fix your issue?
please see https://github.com/openwrt/openwrt/pull/9252#issuecomment-1046299984

**Comment by kestrel1974** [comment]
@abajk @hauke @jospezial I was wrong, I checked again and the gswip_mdio_w method used in the gswip driver do not interfere with the mdio writes exposed through gswip_mdio_xxx functions. So sorry for the confusion I might have caused.
